internal/app/machined/pkg/runtime/v1alpha1/platform/aws/metadata.go

```diff
@@ -0,0 +1,76 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+// MetadataConfig represents a metadata AWS instance.
+type MetadataConfig struct {
+	Hostname     string `json:"hostname,omitempty"`
+	InstanceID   string `json:"instance-id,omitempty"`
+	InstanceType string `json:"instance-type,omitempty"`
+	PublicIPv4   string `json:"public-ipv4,omitempty"`
+	PublicIPv6   string `json:"ipv6,omitempty"`
+	Region       string `json:"region,omitempty"`
+	Zone         string `json:"zone,omitempty"`
+}
+
+//nolint:gocyclo
+func (a *AWS) getMatadata(ctx context.Context) (err error) {
+	getMetadataKey := func(key string) (v string, err error) {
+		v, err = a.metadataClient.GetMetadataWithContext(ctx, key)
+		if err != nil {
+			if awsErr, ok := err.(awserr.RequestFailure); ok {
+				if awsErr.StatusCode() == http.StatusNotFound {
+					return "", nil
+				}
+			}
+
+			return "", fmt.Errorf("failed to fetch %q from IMDS: %w", key, err)
+		}
+
+		return v, nil
+	}
+
+	if a.metadata == nil {
+		a.metadata = &MetadataConfig{}
```

Comment:
this feels like a potential race-condition as we're accessing local variable `a.metadata` from various functions

I'm not sure why we even need it? can't we pull the necessary metadata "twice" if we need to?